src/vivarium_public_health/risks/effect.py

```diff
@@ -316,3 +321,185 @@ def _get_risk_exposure_class(self, builder: Builder) -> Risk:
+
+        # check that rr_data has 1000 parameter values
+        exposure_values = rr_data["parameter"].values
+        if len(np.unique(exposure_values)) != NUM_RR_EXPOSURE_VALUES:
+            raise ValueError(
+                f"The parameter column in your {self.risk.name} relative risk data must contain {NUM_RR_EXPOSURE_VALUES} values to be used in NonLogLinearRiskEffect."
+            )
+
+        # and that these values are monotonically increasing within each demographic group
+        demographic_cols = [
+            col for col in rr_data.columns if col != "parameter" and col != "value"
+        ]
+
+        def values_are_monotonically_increasing(df: pd.DataFrame) -> bool:
+            return np.all(df["parameter"].values[1:] >= df["parameter"].values[:-1])
+
+        group_is_increasing = rr_data.groupby(demographic_cols).apply(
+            values_are_monotonically_increasing
+        )
+        if not group_is_increasing.all():
+            raise ValueError(
+                "The parameter column in your relative risk data must be monotonically increasing to be used in NonLogLinearRiskEffect."
+            )
+
+        def define_rr_intervals(df: pd.DataFrame) -> pd.DataFrame:
+            # create new row for right-most exposure bin (RR is same as max RR)
+            max_exposure_row = df.tail(1).copy()
+            max_exposure_row["parameter"] = np.inf
+            rr_data = pd.concat([df, max_exposure_row]).reset_index()
+
+            rr_data["left_exposure"] = [0] + rr_data["parameter"][:-1].tolist()
+            rr_data["left_rr"] = [rr_data["value"].min()] + rr_data["value"][:-1].tolist()
+            rr_data["right_exposure"] = rr_data["parameter"]
+            rr_data["right_rr"] = rr_data["value"]
+
+            return rr_data[
+                ["parameter", "left_exposure", "left_rr", "right_exposure", "right_rr"]
+            ]
+
+        # define exposure and rr interval columns
+        demographic_cols = [
+            col for col in rr_data.columns if col != "parameter" and col != "value"
+        ]
+        rr_data = (
+            rr_data.groupby(demographic_cols)
+            .apply(define_rr_intervals)
+            .reset_index(level=-1, drop=True)
+            .reset_index()
+        )
+        rr_data = rr_data.drop("parameter", axis=1)
+        rr_data[f"{self.risk.name}_exposure_start"] = rr_data["left_exposure"]
+        rr_data[f"{self.risk.name}_exposure_end"] = rr_data["right_exposure"]
+        # build lookup table
+        rr_value_cols = ["left_exposure", "left_rr", "right_exposure", "right_rr"]
+        self.lookup_tables["relative_risk"] = self.build_lookup_table(
+            builder, rr_data, rr_value_cols
+        )
+
+        paf_data = self.get_filtered_data(
+            builder, self.configuration.data_sources.population_attributable_fraction
+        )
+        self.lookup_tables["population_attributable_fraction"] = self.build_lookup_table(
+            builder, paf_data
+        )
```

Comment:
Build all lookup tables method is really important especially when you are making a complicated component like this for the versatility it offers. It would be a lot easier to read if you made these local functions private class methods.. This would also make user's lives easier if they needed to subclass this component in the future.

Reply:
You also did this with the relative risk data which I think you should rename to _get_relative_risk_data since that is only being used for this class I believe and there is a different method that gets the same data elsewhere (or at least there use to be before a bunch of this risk work).

Reply:
I split out the validation into its own function (which is also what happens in distributions), but I think this component is specific enough that I don't see it getting subclassed. Most likely get_relative_risk_data would need to change anyways, which is already overwriting a function inherited from the parent class.